pkg/metallb/frr.go

```diff
@@ -0,0 +1,589 @@
+}
+
+// Delete removes FrrConfigurationBuilder object from a cluster.
+func (builder *FrrConfigurationBuilder) Delete() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Deleting the FrrConfiguration object %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace,
+	)
+
+	if !builder.Exists() {
+		glog.V(100).Infof(
+			"frrConfiguration %s in namespace %s does not exist",
+			builder.Definition.Name, builder.Definition.Namespace)
+
+		builder.Object = nil
+
+		return nil
+	}
+
+	err := builder.apiClient.Delete(context.TODO(), builder.Definition)
+
+	if err != nil {
+		return fmt.Errorf("can not delete frrConfiguration: %w", err)
+	}
+
+	builder.Object = nil
+
+	return nil
+}
+
+// GetFrrConfigurationGVR returns FrrConfiguration GroupVersionResource which could be used for Clean function.
+func GetFrrConfigurationGVR() schema.GroupVersionResource {
+	return schema.GroupVersionResource{
+		Group: FRRAPIGroup, Version: APIVersion, Resource: "frrconfigurations",
+	}
+}
+
+// Exists checks whether the given FrrConfiguration exists.
+func (builder *FrrConfigurationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if FrrConfiguration %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithBGPRouter defines a BGP router as localASN.
+func (builder *FrrConfigurationBuilder) WithBGPRouter(localASN uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Define BGP router with an ASN: %v", localASN)
+
+	// If there are routers, append a new one
+	builder.Definition.Spec.BGP.Routers = append(builder.Definition.Spec.BGP.Routers, frrtypes.Router{
+		ASN: localASN,
+	})
+
+	return builder
+}
+
+// WithToReceiveModeAll allows all incoming routes.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeAll(routerIndex, neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```

Comment:
This can be removed if we short-circuit return builder.

Reply:
removed